### PR TITLE
Show count of beatmaps in collections in manage dialog

### PR DESCRIPTION
**Pull request by peppy**
Touched on in https://github.com/ppy/osu/discussions/22930

osu! 2023-03-22 at 05 47 27

### CODE REVIEW

File: osu.Game/Collections/DrawableCollectionListItem.cs [hunk 1]
```diff
@@ -117,11 +116,67 @@ private partial class ItemTextBox : OsuTextBox
...
+                    {
+                        Anchor = Anchor.BottomLeft,
+                        Origin = Anchor.TopLeft,
+                        Depth = float.MinValue,
+                        Font = OsuFont.Default.With(size: count_text_size, weight: FontWeight.SemiBold),
+                        Margin = new MarginPadding { Top = 2, Left = 2 },
+                        Colour = colours.Yellow
+                    });
+
+                    itemCountSubscription = realm.SubscribeToPropertyChanged(r => r.Find<BeatmapCollection>(collection.ID), c => c.BeatmapMD5Hashes, _ =>
```
Comment (bdach) [inline osu.Game/Collections/DrawableCollectionListItem.cs:158]: This is instant alarm bells for me. One subscription for each item on the list? So potentially N subscriptions for N items? There's no pooling or list virtualisation anywhere here, so that means one subscription for _every_ collection in db. It's also registered at `OsuGame` level, which means that the disposal of the subscription isn't going to save us any as it just _doesn't fire until the game is closed_.

I fear this is going to fall over _badly_ for power users with monstrous beatmap collections.

- Do we need this to be a subscription, even? The manage collections dialog is modal. There's not really any way for the counts to change while it's open except for background imports - but maybe it's fine not to support that?
- If we do need this to be a subscription, it should probably be handled a lot more economically. Maybe by eagerly acquiring on show and eagerly disposing on hide. Or by pooling/DLUW.

Comment (peppy) [reply osu.Game/Collections/DrawableCollectionListItem.cs:158]: It should be pretty efficient with https://github.com/realm/realm-dotnet/pull/3121, which is why I went for this.

I'd be more interested in moving it once we see it having an adverse effect. And definitely not by pooling (maybe one subscription at the dialog level, although that may have higher overhead due to no keypath filtering).

> Do we need this to be a subscription, even? The manage collections dialog is modal. There's not really any way for the counts to change while it's open except for background imports - but maybe it's fine not to support that?

It's modal, but unless the counts are refreshed each time it's shown, there's plenty of other ways the counts can change (right click menu at song select; collections dropdown at song select).

File: osu.Game/Collections/DrawableCollectionListItem.cs [hunk 2]
```diff
@@ -117,11 +116,67 @@ private partial class ItemTextBox : OsuTextBox
...
+                        Scheduler.AddOnce(() =>
+                        {
+                            int count = collection.PerformRead(c => c.BeatmapMD5Hashes.Count);
+
+                            countText.Text = count == 1
+                                // Intentionally not localised until we have proper support for this (see https://github.com/ppy/osu-framework/pull/4918
+                                // but also in this case we want support for formatting a number within a string).
+                                ? $"{count:#,0} beatmap"
+                                : $"{count:#,0} beatmaps";
```
Comment (Joehuu) [inline osu.Game/Collections/DrawableCollectionListItem.cs:167]: As discussed in https://github.com/ppy/osu/pull/23000, "beatmaps" is incorrect here. I would suggest "items" if we going with the same direction here.

File: osu.Game/Collections/DrawableCollectionListItem.cs (hunk 1 above)
Comment (Joehuu) [reply osu.Game/Collections/DrawableCollectionListItem.cs:158]: Is the supposed performance concern still blocking this? I thought the flow was pushing stuff until someone complained.

Comment (bdach) [reply osu.Game/Collections/DrawableCollectionListItem.cs:158]: I'm not merging without testing. Someone linked a database with a bunch of huge collections so I might give this a try using that and see how hard it dies.

Comment (bdach) [reply osu.Game/Collections/DrawableCollectionListItem.cs:165]: Well I tested using database from https://github.com/ppy/osu/issues/27322#issuecomment-1960943814, which has ~80 collections. Opening the collections dialog increases the number of active subscriptions five-fold. I didn't see signs of falling over, surprisingly - even on mutations - but as predicted the subscriptions stay alive for the entirety of the remaining gameplay and the only thing that gets rid of them is deleting the collections. And 80 is probably not even that many collections.

I dunno, I'm still too skeptical to feel confident pushing this to users. It just feels way too shoddy to be keeping these subscriptions around when the dialog isn't even visible...

**Comment by bdach** [comment]
I've wanted to revive this for a while now that realm has become unstuck. In trying to apply the realm team guidance [linked above](https://github.com/ppy/osu/pull/22932#discussion_r1506351113), I realised that... there is actually no need to add _any_ extra subscriptions to make this work over what's already in `master`. Thus, my humble proposal - along with the explanation as to why no extra subscriptions are required - is https://github.com/ppy/osu/pull/22932/commits/884fa20b286264482f6e965f946369e42d9fe356.

Additionally, I removed the contentious 'beatmaps' terminology in https://github.com/ppy/osu/pull/22932/commits/d8b3c28c2e5cb3c666ae937d4cd13feb7d5475d4.

@peppy is this something you'd be happy with? I can't imagine having (count of user collections) less active at all times for literally free to be a bad thing, but I figure I should double-check.

**Comment by peppy** [comment]
I guess this does work for now yeah.

Although it also shows items as including items which aren't present in the local database (in other words not downloaded). Do you think we should somehow show this, something like "12 items (10 downloaded)"? Or maybe "10 items (+2 not downloaded)"?

Or do we just turn a blind eye?


https://github.com/user-attachments/assets/a277727c-3cff-44cb-83fc-8cc91b68e787

**Comment by bdach** [comment]
> Do you think we should somehow show this, something like "12 items (10 downloaded)"? Or maybe "10 items (+2 not downloaded)"?

I'm a little worried that query will kill things performance-wise, although `BeatmapInfo.MD5Hash` _is_ `[Indexed]`... I'm sorta tempted to just let this slide until someone notices, but I guess we could try and see if it kills performance.

**Comment by peppy** [comment]
I'm also fine with leaving it until someone brings it up, but I think that might happen quite fast (it was the first thing I noticed during testing)

**Comment by bdach** [comment]
Yeah I'd just merge at this point... except I'm a bit more worried about the test failures here.

**Comment by peppy** [comment]
I believe this would fix it

```diff
diff --git a/osu.Game/Collections/DrawableCollectionListItem.cs b/osu.Game/Collections/DrawableCollectionListItem.cs
index f2b00004e2..b0dd70227c 100644
--- a/osu.Game/Collections/DrawableCollectionListItem.cs
+++ b/osu.Game/Collections/DrawableCollectionListItem.cs
@@ -255,7 +255,7 @@ protected override bool OnClick(ClickEvent e)
             private void deleteCollection() => collection.PerformWrite(c => c.Realm!.Remove(c));
         }
 
-        public IEnumerable<LocalisableString> FilterTerms => [(LocalisableString)Model.Value.Name];
+        public IEnumerable<LocalisableString> FilterTerms => Model.PerformRead(m => m.IsValid ? new[] { (LocalisableString)m.Name } : []);
 
         private bool matchingFilter = true;
 

```

tl;dr because collections aren't soft delete, we need to check on every access. doubt it's related to this PR, just very good luck?

**Comment by bdach** [comment]
Yeah my conclusion is similar. Unless something horrendous happens on CI I'll just merge this.

**Review by bdach — APPROVED** [review]
